### ISSUE TITLE
Connect with user credentials?

### ISSUE THREAD

**Issue by ronnyek** [rebus-org/Rebus.RabbitMq#8]
So I'm just starting to tinker with rebus, and noticed that with my message bus of choice (rabbit) it would appear that my only way of providing user/pass is via the amqp connection string. In providing the connection string as ```amqp://user:pass@rabbitaddress``` I get an exception complaining ACCESS_REFUSED because authentication mechanism was PLAIN. 

Is there a way to specify user/pass without doing so via amqp connection string? I'd prefer to actually not send via plain.

**Comment by mookid8000** [comment]
Hmm, weird... while it might be possible to set the credentials on the `ConnectionFactory`, I would expect the client to deliver them to RabbitMQ by connecting to a `user:pass@`-style address.

I am not familiar with this particular error message, but somehow it reminds me of WCF in the old days, which would refuse to do Basic HTTP Auth over a non-encrypted connection (i.e. it would require HTTPS).

Could you try adding an `s` there in the protocol part of the URL, like so:

    amqps://user:pass@rabbitaddress

and see if it changes anything?

**Comment by ronnyek** [comment]
I just played with that making sure I actually enabled the tls/ssl stuff.... I still just keep getting 
```
AuthenticationFailureException: ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.
```

The rabbitmq is hosted on another server... so it may infact be related, but I've successfully connected with the plain .net rabbitmq.client where you just add user and pass to the connection factory.

**Comment by mookid8000** [comment]
Hmm.... I seem to recall I saw somewhere that this particular error message showed when some guy connected to the wrong vhost... are you using anything non-default? (i.e. not '/')?

**Comment by gtresoldi** [comment]
+1 
i use default vhost

**Comment by mookid8000** [comment]
hmmmmm that's weird!

> PLAIN login refused: user 'guest' can only connect via localhost

seems to indicate that the user from the connection string is somehow not correctly passed to the server...

are you using the latest Rebus.RabbitMq package 4.1.0 ? if so, could one of you maybe try and downgrade it to 4.0.0 and see if it works then?

**Comment by gtresoldi** [comment]
In 4.0.1 works fine.
The error begin in 4.1.0

**Comment by mookid8000** [comment]
damn!

The difference between 4.0.0 and 4.1.0 is that 4.1.0 uses the RabbitMQ C# client's built-in failover capabilities, whereas in versions < 4.1.0 Rebus would manage a list of `ConnectionFactory` instances and implement the failover by itself.

@samartzidis Does this issue ring any bells with you?

**Comment by samartzidis** [comment]
Yes I can reproduce this behaviour in regards to credentials in 4.1.0. Will provide a fix ASAP. Thank you.

**Comment by samartzidis** [comment]
I opened a pull request for this. It seems that the underlying RabbitMQ library was not taking into account the credentials in the AMQP connection URI(s) specified in the CreateConnection( ) and needed them to be explicitly provided to the ConnectionFactory. So in the afforementioned faults it was basically trying to connect using the default guest account, ignoring explicit credentials.

**Comment by mookid8000** [comment]
@samartzidis 's fix is out in 4.1.1 now, so if you could try it out and see it the issue has been fixed, that would be great

![giphy 1](https://user-images.githubusercontent.com/32968/32133580-127677b6-bbdb-11e7-946d-ec6f3542422f.gif)

**Comment by mookid8000** [comment]
Excellent! Big thanks to @samartzidis for the quick fix :+1:
